### PR TITLE
US-410084-2:Deployment changes for Client Cloud SRS-ES AuthN

### PR DESCRIPTION
**Pull request by Anand-Kotamsetti**
1)Added Basic Authentication and AWS IAM Role authentication support for SRS

### CODE REVIEW

File: charts/backingservices/values.yaml [hunk 1]
```diff
@@ -29,10 +29,18 @@ srs:
     # To use your own Elasticsearch cluster, set srsStorage.provisionInternalESCluster to false and then
     # set the external Elasticsearch cluster URL and port details below when using an externally managed elasticsearch
     # Ensure that the specified endpoint is accessible from the kubernetes cluster pods.
-    # Currently the elasticsearch connection does not support any modes of authentication and should be es endpoint APIs' accessible without authentication.
+    # Currently the elasticsearch connection is supported with Basic Authentication  and AWS IAM Role based Authentication .
 #    domain: managed-elasticsearch.acme.io
 #    port: 443
 #    protocol: https
+    # If externally managed Elasticsearch cluster is authenticated with Basic Authentication then please set below srs.srsStorage.basicAuthentication's srs.srsStorage.basicAuthentication.username and srs.srsStorage.basicAuthentication.password values
```
Comment (taz-pega-work) [inline charts/backingservices/values.yaml:36]: To configure authentication for the externally managed Elasticsearch cluster to use Basic Authentication then uncomment and add the parameters, srs.srsStorage.basicAuthentication.username and srs.srsStorage.basicAuthentication.password

To configure authentication for the externally managed Elasticsearch cluster to use an AWS IAM Role then uncomment and add the parameters, srs.srsStorage.awsIAM and srs.srsStorage.awsIAM.region

File: charts/backingservices/values.yaml [hunk 2]
```diff
@@ -47,6 +55,24 @@ elasticsearch:
   # Shrink default JVM heap.
   esJavaOpts: "-Xmx1024m -Xms1024m"
   # Allocate smaller chunks of memory per pod.
+  protocol: http
+  # This section specifies the elasticsearch cluster configuration for authentication.
+  esConfig:
+    elasticsearch.yml: |
+      xpack.security.enabled: true
+      xpack.security.transport.ssl.enabled: true
+  #Extra Environmental Variables for elasticsearch basic authentication , Values are read from Secrets
```
Comment (taz-pega-work) [inline charts/backingservices/values.yaml:64]: Use this section to include additional, supported environmental variables for Elasticsearch basic authentication. The parameter values can be read from a specified secrets file.

**Review by taz-pega-work — COMMENTED** [review]
I added a little more clarity to a few of your comments. In general these are helpful comments and ready to go from my perspective.

File: charts/backingservices/charts/srs/templates/_helpers.tpl [hunk 3]
```diff
@@ -125,6 +125,45 @@ app.kubernetes.io/instance: {{ .Release.Name }}
 {{- end }}
 {{- end }}
 
+{{- define "elasticsearch.authProvider" -}}
+{{- if (.Values.srsStorage.provisionInternalESCluster) -}}
+{{- "basic-authentication" }}
+    {{- else }}
```
Comment (MadhuriArugula) [inline charts/backingservices/charts/srs/templates/_helpers.tpl:131]: Can we have this else block inline to `if` to get a clear understanding of code.

File: charts/backingservices/charts/srs/templates/_helpers.tpl [hunk 4]
```diff
@@ -125,6 +125,45 @@ app.kubernetes.io/instance: {{ .Release.Name }}
...
+        {{- if and  (.Values.srsStorage.basicAuthentication) (not .Values.srsStorage.awsIAM ) -}}
+        {{- "basic-authentication" }}
+        {{- else if and  (.Values.srsStorage.awsIAM)  (not .Values.srsStorage.basicAuthentication ) -}}
+        {{- "aws-iam"}}
+        {{- else if and  (not .Values.srsStorage.basicAuthentication ) (not .Values.srsStorage.awsIAM )}}
+        {{- "none" }}
+        {{- else if and ( .Values.srsStorage.basicAuthentication ) ( .Values.srsStorage.awsIAM )}}
+        {{- fail "Only one authentication can be enabled, please try to disable .Values.srsStorage.basicAuthentication/.Values.srsStorage.awsIAM when .Values.srsStorage.provisionInternalESCluster is false" | quote  }}
+        {{- else}}
+        {{- fail "Please disable/comment Values.srsStorage.basicAuthentication or Values.srsStorage.awsIAM when '.Values.srsStorage.provisionInternalESCluster' is true" | quote  }}
```
Comment (MadhuriArugula) [inline charts/backingservices/charts/srs/templates/_helpers.tpl:141]: Is this condition reachable? If not, can we remove this `else` statement?

**Review by MadhuriArugula — CHANGES_REQUESTED** [review]

**Review by MadhuriArugula — APPROVED** [review]

**Review by Saurabh-16 — APPROVED** [review]

File: charts/backingservices/values.yaml [hunk 5]
```diff
@@ -29,11 +29,22 @@ srs:
...
+#     To configure authentication for the externally managed Elasticsearch cluster to use Basic Authentication then uncomment
+#     add the parameters, srs.srsStorage.basicAuthentication.username and srs.srsStorage.basicAuthentication.password
+#    basicAuthentication:
+#      username: "BASIC_AUTH_USERNAME"
+#      password: "BASIC_AUTH_PASSWORD"
+#     To configure authentication for the externally managed Elasticsearch cluster to use an AWS IAM Role then uncomment and
+#     add the parameters, srs.srsStorage.awsIAM and srs.srsStorage.awsIAM.region
+#     srs.srsStorage.awsIAM's srs.srsStorage.awsIAM.region value
+#    awsIAM:
+#      region: "AWS_ELASTICSEARCH_REGION"
```
[the remark this answers is not in the document]
Comment (Anand-Kotamsetti) [reply charts/backingservices/values.yaml:45]: Thanks for feedback,
We do have similar flow to identify the configuration value in SRS. As this configuration value is internal to SRS we are not expecting this values to be passed as input by client.

File: charts/backingservices/values.yaml [hunk 6]
```diff
@@ -29,11 +29,22 @@ srs:
     # To use your own Elasticsearch cluster, set srsStorage.provisionInternalESCluster to false and then
     # set the external Elasticsearch cluster URL and port details below when using an externally managed elasticsearch
     # Ensure that the specified endpoint is accessible from the kubernetes cluster pods.
-    # Currently the elasticsearch connection does not support any modes of authentication and should be es endpoint APIs' accessible without authentication.
+    # Currently the elasticsearch connection is supported with Basic Authentication  and AWS IAM Role based Authentication .
```
Comment (taz-pega-work) [inline charts/backingservices/values.yaml:32]: The elasticsearch connection supports two authentication methods: basic authentication or AWS IAM role-based authentication.

File: charts/backingservices/values.yaml [hunk 7]
```diff
@@ -29,11 +29,22 @@ srs:
...
 #    domain: managed-elasticsearch.acme.io
 #    port: 443
 #    protocol: https
-#    set `requireInternetAccess` to true when the elasticsearch domain is outside of the Kubernetes cluster network  and is available over internet
+#     To configure authentication for the externally managed Elasticsearch cluster to use Basic Authentication then uncomment
```
Comment (taz-pega-work) [inline charts/backingservices/values.yaml:36]: To configure basic authentication to your externally-managed Elasticsearch cluster,  uncomment and add the parameter details: srs.srsStorage.basicAuthentication.username and srs.srsStorage.basicAuthentication.password

File: charts/backingservices/values.yaml [hunk 8]
```diff
@@ -29,11 +29,22 @@ srs:
...
+#     add the parameters, srs.srsStorage.basicAuthentication.username and srs.srsStorage.basicAuthentication.password
+#    basicAuthentication:
+#      username: "BASIC_AUTH_USERNAME"
+#      password: "BASIC_AUTH_PASSWORD"
+#     To configure authentication for the externally managed Elasticsearch cluster to use an AWS IAM Role then uncomment and
```
Comment (taz-pega-work) [inline charts/backingservices/values.yaml:41]: To configure AWS IAM role-based authentication to your externally-managed Elasticsearch cluster, uncomment and add the parameter details:  srs.srsStorage.awsIAM and its associated region, srs.srsStorage.awsIAM.region

File: charts/backingservices/values.yaml [hunk 9]
```diff
@@ -29,11 +29,22 @@ srs:
...
+#     add the parameters, srs.srsStorage.awsIAM and srs.srsStorage.awsIAM.region
+#     srs.srsStorage.awsIAM's srs.srsStorage.awsIAM.region value
+#    awsIAM:
+#      region: "AWS_ELASTICSEARCH_REGION"
+
+#     set `requireInternetAccess` to true when the elasticsearch domain is outside of the Kubernetes cluster network  and is available over internet
```
Comment (taz-pega-work) [inline charts/backingservices/values.yaml:47]: To configure either authentication method, when the elasticsearch domain requires an open internet connection, uncomment to set this parameter to "true".

**Review by taz-pega-work — COMMENTED** [review]
The in-line guidance has inconsistent punctuation and is hard to read, especially hte IAM role-bases settings...don't add possessive "'s" to a parameter name...it's very confusing ;-)  ...I added suggestions to clarify instructions for your users.

**Comment by taz-pega-work** [comment]
documentation content is good!

**Review by kishorv10 — APPROVED** [review]

**Review by pega-kshev — APPROVED** [review]

**Review by vanga — APPROVED** [review]

**Review by pega-kshev — APPROVED** [review]

**Review by pega-kshev — APPROVED** [review]

**Review by pega-kshev — APPROVED** [review]

**Review by taz-pega-work — APPROVED** [review]
looks good. thanks for making changes in the runbooks!

**Pull request merged**